### DISCUSSION TITLE
Very slow to import Client...

### DISCUSSION THREAD

**Discussion by edwardsnj** [MarshalX/atproto#441] [Questions] [answered]
Importing from atproto seems to take a long time. I'm on raspios bullseye, python 3.9. atproto v0.0.55. Yes, this is a low-powered place to run this, but this is extreme slowness. 

Installed in ~/.local/... to ensure modules can be compiled etc... but this doesn't seem to help

% echo "from atproto import Client" | time -p python3
real 123.54
user 104.61
sys 1.20
%  echo "from atproto import Client" | time -p python3
real 108.29
user 105.61
sys 0.98
%  echo "from atproto import Client" | time -p python3
real 103.41
user 100.71
sys 1.21

Even going directly to the module doesn't seem to help:

% echo "from atproto_client.client.client import Client" | time -p python3
real 112.46
user 110.25
sys 1.08

What is it doing for 100+ seconds? How can I speed this up?

**Comment by MarshalX** [comment]
Wow... it takes up to 1 sec to me, but 100 is insane. All goes from this issue: https://github.com/pydantic/pydantic/issues/9908

Now imagine that we have 400+ pydatic models

**Comment by edwardsnj** [comment]
OK, I have a hypothesis, will take me some time to test it. I'm guessing that the pydantic_core module (written in Rust, to speed it up) does not have a pre-compiled 32-bit ARM version of the module available. I'm on a Raspberry Pi Zero W. So it may be falling back to a pure python strategy, which is slooooow. I'm going to try to compile pydantic_core myself and see if that helps.

**Comment by edwardsnj** [comment]
As far as I can tell, my hypothesis is not correct - there is a compiled _pydantic_core module: _pydantic_core.cpython-39-arm-linux-gnueabihf.so and it appears to be loaded correctly. So its just very slow. While I love how quickly I was able to make a bluesky post client in python using this module, maybe its too heavyweight for my needs. I wonder what alternatives I have in python. I guess I can roll my own that just hits the web-service API to post.

**Reply by MarshalX** [reply]
That's very sad news 😞

**Reply by antont** [reply]
> I wonder what alternatives I have in python. I guess I can roll my own that just hits the web-service API to post.

David Buchanan is making a minimalistic PDS in Python & C, I don't know if it could be helpful but maybe he has a lightweight atproto impl there (i didn't look) https://github.com/DavidBuchanan314/millipds

**Reply by antont** [reply]
apparently he uses a couple of his own libs for the AT stuff and the CBR code is indeed in C there https://github.com/DavidBuchanan314/dag-cbrrr/blob/main/src/cbrrr/_cbrrr.c

**Comment by MarshalX** [comment]
pydantic getting serious about import time now! they reported significant improvements in the new alpha version. feel free to test. more info: https://github.com/pydantic/pydantic/issues/9908#issuecomment-2625417484

**Comment by MarshalX** [comment, marked as answer]
pydantic team did a very good job in the v2.11 release. the importing time and memory usage have been improved! 

- their release with details: https://pydantic.dev/articles/pydantic-v2-11-release
- my comparison with models of atproto and bluesky: https://github.com/MarshalX/pydantic-memory-usage/commit/f3f2d3600ba1148d4abbf8961a128f028c8fdb9a

i highly recommend updating pydantic; gonna mark this as answered

**Comment by edwardsnj** [comment]
For my limited use-case, writing a bare-bones poster that hit the API directly was sufficient. Feel free to close the ticket.
